### DISCUSSION TITLE
Can we have common test utilities when running with `--import-mode=importlib`?

### DISCUSSION THREAD

**Discussion by pmeier** [pytest-dev/pytest#9109] [Q&A] [answered]
The [documentation for `--import-mode=importlib` states](https://docs.pytest.org/en/6.2.x/pythonpath.html#import-modes):

> For this reason this [...] makes test modules non-importable by each other. [...] Users which require this should turn their tests into proper packages instead.

For me this sounds like if the tests directory is a proper package, we can import from other modules within it. Unfortunately, I didn't get it working:

```
tests/
├── __init__.py
├── test_foo.py
└── utils.py
```

with `test_foo.py`:

```python
from tests.utils import bar

def test_foo():
    pass
```

Running `pytest --import-mode=importlib --co tests/` results in 

```
E   ModuleNotFoundError: No module named 'tests'
```

I get similar errors if I change the import to `from .utils import bar` or `from utils import bar`.

Did I do something wrong in the setup or is the documentation misleading and importing from other test utilities is not supported for `--import-mode=importlib`?

**Comment by nicoddemus** [comment, marked as answer]
Hi @pmeier, 

> Did I do something wrong in the setup or is the documentation misleading and importing from other test utilities is not supported for --import-mode=importlib?

The latter is true, `--import-mode=importlib` will import your tests without modifying `sys.path` at all, which means tests cannot import from each other.

There's https://github.com/pytest-dev/pytest/issues/8964 about adding some support for that, but that discussion lost steam.

**Reply by pmeier** [reply]
That is unfortunate, but thanks a lot for the pointer. I'll comment on the issue.
